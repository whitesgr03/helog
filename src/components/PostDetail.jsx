import style from "../styles/PostDetail.module.css";
import { imageWrap } from "../styles/image.module.css";

import PropTypes from "prop-types";
import { format } from "date-fns";

const post = {
	id: "1",
	url: "#",
	title: "This is title A",
	content: "This is content A",
	createdAt: new Date("2024/5/1"),
	lastModified: new Date("2024/5/1"),
};

const PostDetail = () => {
	return (
		<ul className={style.postDetail}>
			<li>
				<h2>{post.title}</h2>
				<strong className={style.dateTime}>
					Published in {format(post.createdAt, "MMMM d, y")}
				</strong>
				{post.createdAt.getTime() !== post.lastModified.getTime() && (
					<em>Edited in {format(post.lastModified, "MMMM d, y")}</em>
				)}
				<div className={imageWrap}>
					<img src={post.url} alt={post.title} />
				</div>
				<p>{post.content}</p>
			</li>
		</ul>
	);
};

PostDetail.propTypes = {
	post: PropTypes.object,
};

export default PostDetail;
