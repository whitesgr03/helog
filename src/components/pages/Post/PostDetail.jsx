// Packages
import { useEffect, useState } from 'react';
import {
	useParams,
	useOutletContext,
	Link,
	Navigate,
	useLocation,
} from 'react-router-dom';
import { format } from 'date-fns';
import { Editor } from '@tinymce/tinymce-react';

// Styles
import styles from './PostDetail.module.css';
import imageStyles from '../../../styles/image.module.css';

// Components
import { Loading } from '../../utils/Loading';
import { Comments } from '../Comment/Comments';

// Utils
import { getPostDetail } from '../../../utils/handlePost';

export const PostDetail = () => {
	const { postId } = useParams();
	const { posts, onUpdatePost } = useOutletContext();
	const [loadContent, setLoadContent] = useState(true);
	const [loading, setLoading] = useState(true);
	const [error, setError] = useState(null);
	const imageContentRef = useRef(null);

	const post = posts.find(post => post._id === postId);

	const { pathname: previousPath } = useLocation();

	useEffect(() => {
		const controller = new AbortController();
		const { signal } = controller;

		const handleGetPostDetail = async () => {
			const result = await getPostDetail({ postId: post._id, signal });
			const handleResult = () => {
				result.success
					? onUpdatePost({ postId: post._id, newPost: result.data })
					: setError(result.message);
				setLoading(false);
			};

			result && handleResult();
		};

		post?.content === undefined ? handleGetPostDetail() : setLoading(false);

		return () => controller.abort();
	}, [post, onUpdatePost]);

	return (
		<>
			{error ? (
				<Navigate to="/error" state={{ error, previousPath }} />
			) : (
				<>
					<div id="post-detail" className={styles['post-detail']}>
						{(loading || loadContent) && <Loading text={'Loading post...'} />}
						<div
							className={`${styles.container} ${loading || loadContent ? styles.loading : ''}`}
						>
							<Link to={-1} className={styles.link}>
								<span
									className={`${styles['left-arrow']} ${imageStyles.icon}`}
								/>
								Back to previous page
							</Link>
							{post?.title && <h2 className={styles.title}>{post.title}</h2>}
							<div className={styles['date-time']}>
								<strong>{post.author.username}</strong>
								<em>{`Published in ${format(post.createdAt, 'MMMM d, y')}`}</em>
								{new Date(post.createdAt).getDate() !==
									new Date(post.updatedAt).getDate() && (
									<em>{`Edited in ${format(post.updatedAt, 'MMMM d, y')}`}</em>
								)}
								{post?.mainImageUrl && (
									<div className={styles['image-wrap']}>
										<div className={imageStyles.content} ref={imageContentRef}>
											<img
												src={post.mainImageUrl}
												alt={`${post.title}'s main image`}
											/>
										</div>
									</div>
								)}
							</div>
							<Editor
								tinymceScriptSrc="/tinymce/tinymce.min.js"
								licenseKey="gpl"
								id="editor"
								disabled={true}
								value={post.content ?? ''}
								onInit={() => {
									setLoadContent(false);
								}}
								init={{
									menubar: false,
									toolbar: false,
									inline: true,
									plugins: 'codesample',
								}}
							/>
						</div>
					</div>
					{!loading && !loadContent && <Comments post={post} />}
				</>
			)}
		</>
	);
};
