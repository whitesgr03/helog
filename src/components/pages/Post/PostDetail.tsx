// Packages
import { useEffect, useState, useRef } from 'react';
import { useParams, Link, Navigate, useLocation } from 'react-router-dom';
import { format } from 'date-fns';
import { Editor } from '@tinymce/tinymce-react';
import { Editor as TinyMCEEditor } from 'tinymce';
import { useQuery } from '@tanstack/react-query';

// Styles
import styles from './PostDetail.module.css';
import imageStyles from '../../../styles/image.module.css';

// Components
import { Loading } from '../../utils/Loading';
import { Comments } from '../Comment/Comments';

// Utils
import { queryPostDetailOption } from '../../../utils/queryOptions';

export const PostDetail = () => {
	const { postId = '' } = useParams() ?? {};
	const [contentEditorLoad, setContentEditorLoad] = useState(true);
	const [checking, setChecking] = useState(true);
	const [errorMainImage, setErrorMainImage] = useState(false);

	const imageContentRef = useRef<HTMLDivElement>(null);
	const editorRef = useRef<TinyMCEEditor | null>(null);

	const { pathname: previousPath } = useLocation();

	const {
		isPending,
		isError,
		data: post,
		error,
	} = useQuery(queryPostDetailOption(postId));

	const { clientWidth = '350', clientHeight = '200' } =
		imageContentRef.current ?? {};

	const errorImageUrl = `https://fakeimg.pl/${clientWidth}x${clientHeight}/?text=404%20Error&font=noto`;

	const handleError = () => setErrorMainImage(true);
	const handleLoad = (e: React.SyntheticEvent<HTMLImageElement>) =>
		(e.currentTarget.width <= 0 || e.currentTarget.height <= 0) &&
		handleError();

	useEffect(() => {
		let editorContent = editorRef.current?.getContent() ?? '';
		const imgs = editorContent?.match(/<img.+">/g) ?? [];

		const handleCheckContentImages = async () => {
			const imageErrors = await Promise.all(
				imgs.map(
					img =>
						new Promise(resolve => {
							const [url = ''] = img.match(/(?<=src=")(.*?)(?=")/g) ?? [];
							const [width = '350', height = '200'] =
								img.match(/(?<=width:\s|height:\s).*?(?=px;)/g) ?? [];

							const image = new Image();
							const handleError = () => {
								const errorImageUrl = `https://fakeimg.pl/${width}x${height}/?text=404%20Image%20Error&font=noto`;

								editorContent = editorContent.replace(url, errorImageUrl);

								resolve(true);
							};
							const handleLoad = () =>
								image.width <= 0 || image.height <= 0
									? handleError()
									: resolve(false);
							image.onerror = handleError;
							image.onload = handleLoad;
							image.src = url;
						}),
				),
			);

			Array.isArray(imageErrors) &&
				imageErrors.find(error => error === true) &&
				editorRef.current?.setContent(editorContent);

			setChecking(false);
		};

		!isPending && !contentEditorLoad && handleCheckContentImages();
	}, [isPending, contentEditorLoad]);

	return (
		<>
			<div id="post-detail" className={styles['post-detail']}>
				{isError ? (
					error.cause.status === 404 ? (
						<Navigate to="/error/404" />
					) : (
						<Navigate
							to="/error"
							state={{
								previousPath,
							}}
						/>
					)
				) : (
					checking && <Loading text={'Loading post ...'} />
				)}

				<div
					className={`${styles.container} ${checking ? styles.hide : ''}`}
					data-testid="container"
				>
					{post && (
						<>
							<Link to="/posts" className={styles.link}>
								<span
									className={`${styles['left-arrow']} ${imageStyles.icon}`}
								/>
								Back to list
							</Link>

							<h2 className={styles.title} title={post.title}>
								{post.title}
							</h2>

							<div className={styles['published-date']}>
								<strong>{post.author.username}</strong>

								<em>{`Published in ${format(post.createdAt, 'MMMM d, y')}`}</em>
								{new Date(post.createdAt).getDate() !==
									new Date(post.updatedAt).getDate() && (
									<em>{`Edited in ${format(post.updatedAt, 'MMMM d, y')}`}</em>
								)}
							</div>

							<div className={styles['image-wrap']}>
								<div className={imageStyles.content} ref={imageContentRef}>
									<img
										title={post.title}
										src={errorMainImage ? errorImageUrl : post.mainImage}
										alt="Main image"
										onError={handleError}
										onLoad={handleLoad}
									/>
								</div>
							</div>
						</>
					)}
					<Editor
						id="editor-content"
						initialValue={post?.content ?? ''}
						onInit={(_evt, editor) => {
							editorRef.current = editor;
							setContentEditorLoad(false);
						}}
						init={{
							menubar: false,
							toolbar: false,
							inline: true,
							plugins: 'codesample',
						}}
					/>
				</div>
			</div>

			{!isPending && !isError && (
				<div className={`${checking ? styles.hide : ''}`}>
					<Comments postId={postId} />
				</div>
			)}
		</>
	);
};
