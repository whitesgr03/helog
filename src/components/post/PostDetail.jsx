// Packages
import { useState, useEffect } from "react";
import { useParams, useOutletContext, Link } from "react-router-dom";
import PropTypes from "prop-types";
import { format } from "date-fns";
import { Editor } from "@tinymce/tinymce-react";

// Styles
import style from "../../styles/post/PostDetail.module.css";
import image from "../../styles/utils/image.module.css";

// Components
import Loading from "../layout/Loading";
import Error from "../layout/Error";
import Comments from "../comment/Comments";

// Utils
import { getPostDetail } from "../../utils/handlePost";

const PostDetail = () => {
	const { postId } = useParams();
	const { user } = useOutletContext();
	const [post, setPost] = useState([]);
	const [error, setError] = useState(null);
	const [loading, setLoading] = useState(true);
	const [loadContent, setLoadContent] = useState(true);

	const createdAt = post?.createdAt && new Date(post.createdAt).getTime();
	const lastModified =
		post?.lastModified && new Date(post.lastModified).getTime();

	useEffect(() => {
		const controller = new AbortController();
		const { signal } = controller;

		const handleGetPostDetail = async () => {
			const result = await getPostDetail({ postId, signal });

			const handleResult = () => {
				result.success
					? setPost(result.data)
					: setError(result.message);
				setLoading(false);
			};

			result && handleResult();
		};

		handleGetPostDetail();
		return () => controller.abort();
	}, [user, postId]);
	return (
		<>
			{error || post?.content === "" ? (
				<Error message={error} />
			) : (
				<>
					<div
						className={`${
							loading || loadContent ? style.loading : ""
						}`}
					>
						<div id="postDetail" className={style.postDetail}>
							<Link to="/posts" className={style.link}>
								<span
									className={`${style.leftArrow} ${image.icon}`}
								/>
								Back to list
							</Link>

							{post?.title && (
								<h2 className={style.title}>{post.title}</h2>
							)}
							<div className={style.dateTime}>
								<strong>{post?.author?.name}</strong>
								{post?.lastModified && (
									<em>
										{`${
											createdAt === lastModified
												? "Published"
												: "Edited"
										} in ${format(
											createdAt === lastModified
												? post.createdAt
												: post.lastModified,
											"MMMM d, y"
										)}`}
									</em>
								)}

								{post?.mainImageUrl && (
									<div className={style.imageWrap}>
										<div className={image.content}>
											<img
												src={post.mainImageUrl}
												alt={`${post.title} main image`}
											/>
										</div>
									</div>
								)}
							</div>
							<Editor
								apiKey="x2zlv8pvui3hofp395wp6my8308b15h3s176scf930dizek1"
								id="content"
								disabled={true}
								value={post?.content}
								onInit={() => {
									setLoadContent(false);
								}}
								init={{
									menubar: false,
									toolbar: false,
									inline: true,
									plugins: "codesample",
								}}
							/>
						</div>
						<Comments
							postAuthorId={post?.author?._id}
							postId={postId}
						/>
					</div>
					{(loading || loadContent) && <Loading />}
				</>
			)}
		</>
	);
};

PostDetail.propTypes = {
	post: PropTypes.object,
};

export default PostDetail;
