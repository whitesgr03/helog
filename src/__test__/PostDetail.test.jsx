import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
// import { format } from "date-fns";
// import { BrowserRouter } from "react-router-dom";

import PostDetail from "../components/PostDetail";

describe("PostDetail component", () => {
	it("should render correctly", () => {
		const { asFragment } = render(<PostDetail />);

		const actual = asFragment();

		expect(actual).toMatchSnapshot();
	});
});
