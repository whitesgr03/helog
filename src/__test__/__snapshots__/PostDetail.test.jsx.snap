// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostDetail component > should render correctly 1`] = `
<DocumentFragment>
  <ul
    class="_postDetail_30e276"
  >
    <li
      class="_item_30e276"
    >
      <h2>
        This is title A
      </h2>
      <strong
        class="_dateTime_30e276"
      >
        Published in May 1, 2024
      </strong>
      <div
        class="_imageWrap_30e276"
      >
        <img
          alt="This is title A"
          src="#"
        />
      </div>
      <p>
        This is content A
      </p>
    </li>
  </ul>
</DocumentFragment>
`;
